src/map/status.c

```diff
@@ -12325,6 +12291,151 @@ bool status_readdb_scconfig(char* fields[], int columns, int current) {
+				continue;
+			
+			int level=0, chance=0;
+			
+			if (!libconfig->setting_lookup_int(tt, "Level", &level) || level <= 0 || level > MAX_REFINE) {
+				ShowError("status_readdb_refine_libconfig_sub: Invalid 'Level' configuration %d in entry #%d of \"%s\".\n", level, n, source);
+				return 0;
+			}  else if (!libconfig->setting_lookup_int(tt, "Chance", &chance)) {
+				ShowWarning("status_readdb_refine_libconfig_sub: Missing 'Chance' configuration in entry #%d of \"%s\", defaulting to 0.\n", n, source);
+				status->dbs->refine_info[type].chance[i] = 0;
+			}
+			
+			level -= 1;
+			
+			if( duplicate[level] ) {
+				ShowWarning("status_readdb_refine_libconfig_sub: duplicate entry of Level %d in Type #%d of \"%s\".\n", level, type, source);
+			} else duplicate[level] = true;
+			
+			if ( chance ) status->dbs->refine_info[type].chance[level] = chance;
+			else status->dbs->refine_info[type].chance[level] = 0;
+		}
+		for(i=0; i < MAX_REFINE; i++) {
+			int bonus=0;
+			tt = libconfig->setting_get_elem(t, i);
+			
+			if (!tt)
+				break;
+			else if (!config_setting_is_group(tt))
+				continue;
+			
+			if ( i >= random_bonus_start_level - 1 )
+				status->dbs->refine_info[type].randombonus_max[i] = random_bonus * (i - random_bonus_start_level + 2);
+			if (!libconfig->setting_lookup_int(tt, "Bonus", &bonus) || !bonus )
+				status->dbs->refine_info[type].bonus[i] = 0;
+			else {
+				status->dbs->refine_info[type].bonus[i] = bonus_per_level + bonus;
+				if ( i > 0 ) status->dbs->refine_info[type].bonus[i] += status->dbs->refine_info[type].bonus[i-1];
+			}
+		}
+	} else {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing Rate configuration in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	}
+	
+	return type+1;
+}
+
+bool status_refinedb_lookup_const(const config_setting_t *r, const char *name, int *value) {
+	if (libconfig->setting_lookup_int(r, name, value)) {
+		return true;
+	} else {
+		const char *str = NULL;
+		if (libconfig->setting_lookup_string(r, name, &str)) {
+			if (*str && script->get_constant(str, value))
+				return true;
+		}
+	}
+	return false;
+}
+
+/**
+ * Reads from a libconfig-formatted refine_db.conf file.
+ *
+ * @param *filename File name, relative to the database path.
+ * @return The number of found entries.
+ */
+int status_readdb_refine_libconfig(const char *filename) {
+	bool duplicate[REFINE_TYPE_MAX];
+	config_t refine_db_conf;
+	config_setting_t *rdb, *r;
+	char filepath[256];
+	int i = 0, count = 0;
+	
+	sprintf(filepath, "%s/%s", map->db_path, filename);
+	memset(&duplicate,0,sizeof(duplicate));
+	if( libconfig->read_file(&refine_db_conf, filepath) || !(rdb = libconfig->setting_get_member(refine_db_conf.root, "refine_db")) ) {
+		ShowError("can't read %s\n", filepath);
+		return 0;
+	}
+	
+	while( (r = libconfig->setting_get_elem(rdb,i++)) ) {
+		int type = status->readdb_refine_libconfig_sub(r, i-1, filename);
+		if( !type )
+			continue;
+		
+		if( duplicate[type] ) {
```

Comment:
You probably meant `type-1` (because `status_readdb_refine_libconfig_sub` returns `type+1`, line 12384 of this diff)